clang/test/SemaSYCL/native_cpu/kernelhandler-scalar.cpp

```diff
@@ -0,0 +1,98 @@
+// RUN: %clangxx -fsycl-device-only  -fsycl-targets=native_cpu -Xclang -fsycl-int-header=%t.h -S -o %t.ll %s 
+// RUN: FileCheck -input-file=%t.h.hc %s 
+// RUN: FileCheck -input-file=%t.ll %s --check-prefix=CHECK-LL
+// Compiling generated main integration header to check correctness, -fsycl option used to find required includes
+// RUN: %clangxx -fsycl -D __SYCL_NATIVE_CPU__ -c -x c++ %t.h
+#include <CL/sycl.hpp>
+
+#include <iostream>
+
+using namespace cl::sycl;
+
+const size_t N = 10;
+
+template <typename T>
+class init_a;
+
+template <typename T>
+bool test(queue myQueue) {
+  {
+    buffer<T, 1> a(range<1>{N});
+    T test = 42;
+
+    myQueue.submit([&](handler& cgh) {
+      auto A = a.template get_access<access::mode::write>(cgh);
+      cgh.parallel_for<init_a<T>>(range<1>{N}, [=](id<1> index) {
+        A[index] = test;
+      });
+    });
+
+    auto A = a.template get_access<access::mode::read>();
+    std::cout << "Result:" << std::endl;
+    for (size_t i = 0; i < N; i++) {
+        if (A[i] != test) {
+          std::cout << "ERROR\n";
+          return false;
+        }
+    }
+  }
+
+  std::cout << "Good computation!" << std::endl;
+  return true;
+}
+
+int main() {
+  queue q;
+  int res1 = test<int>(q);
+  int res2 = test<unsigned>(q);
+  int res3 = test<float>(q);
+  int res4 = test<double>(q);
+  if(!(res1 && res2 && res3 && res4)) {
+    return 1;
+  }
+  return 0;
+}
+
+
+
+// CHECK:extern "C" void _Z6init_aIiE(void *, void *, int, __nativecpu_state *);
```

Comment:
How is this being generated? Am I understanding correctly that this is being generated by some backend pass? If so, I do not know if this should be a front end test. Frontend tests usually do not invoke the driver as you have done in these tests. We use clang_cc1 and mock headers in SemaSYCL and CodeGenSYCL.

Reply:
Yes, this header is generated by the `EmitSYCLNativeCPUHeader` pass (part of this patch) because the optimizer may remove unused kernel parameters, which happens frequently in SYCL kernels especially if they capture complex objects. We wanted this to be a frontend test to ensure the correct kernel subhandlers (stubs that call the actual kernel) are generated for certain SYCL kernels.

Reply:
This tests are checking C++ code (from our integration header) that gets emitted by LLVM passes, as @uwedolinsky said we have to emit that code in an LLVM pass because we need to know which kernel arguments have been optimized out. I agree that `SemaSYCL` is not the right place for this tests, I think I originally put them there because the main SYCL integration header is emitted by `Sema`, but I'm happy to move them where you think it's more appropriate, I'll also try using the mock headers, thank you

Reply:
I'm not sure where the test should live because I do not know where tests checking output from backend passes should live. Maybe @AlexeySachkov will know?

IMO the only FE tests that should accompany this change are ones checking `kernel_arg_type` metadata for NativeCPU, a test testing kernel name mangling and possibly the new information emitted in integration footer. All these tests should use clang_cc1 and mock headers

Reply:
The mock headers seem to duplicate a lot of code from the main SYCL headers, and need to be kept consistent with them which is error-prone and could be a lot of work, especially when integrating new changes/updates from upstream LLVM. So maybe going forward it might be better for our `nativecpu`  target to put new tests into the `test-e2e` folder instead so that the tests use the real SYCL headers and the maintenance effort of mock headers is limited/contained?

Reply:
I've moved the tests we add in `SemaSYCL` to `sycl/check_device_code`, I decided to move them there because I couldn't test the same behaviour using mock headers. I've added a smaller test that uses the `cc1` invocation and mock headers in `CodeGenSYCL`, which checks just kernel name mangling, `kernel_arg_type` metadata node and the Native CPU kernel flag. I'm not sure if `CodeGenSYCL` is the right place for that test, please let me know :)

Reply:
Thanks!